aya/src/programs/extension.rs

```diff
@@ -0,0 +1,187 @@
+/// Ok::<(), aya::BpfError>(())
+/// ```
+#[derive(Debug)]
+#[doc(alias = "BPF_PROG_TYPE_EXT")]
+pub struct Extension {
+    pub(crate) data: ProgramData,
+}
+
+impl Extension {
+    /// Loads the program inside the kernel.
+    ///
+    /// See also [`Program::load`](crate::programs::Program::load).
+    pub fn load<T: AsRawFd>(&mut self, program: T, func_name: &str) -> Result<(), ProgramError> {
+        let target_prog_fd = program.as_raw_fd();
+
+        let info = sys::bpf_obj_get_info_by_fd(target_prog_fd).map_err(|io_error| {
+            ProgramError::SyscallError {
+                call: "bpf_obj_get_info_by_fd".to_owned(),
+                io_error,
+            }
+        })?;
+
+        if info.btf_id == 0 {
+            return Err(ProgramError::ExtensionError(ExtensionError::NoBTF));
+        }
+
+        let btf_fd = sys::bpf_btf_get_fd_by_id(info.btf_id).map_err(|io_error| {
+            ProgramError::SyscallError {
+                call: "bpf_btf_get_fd_by_id".to_owned(),
+                io_error,
+            }
+        })?;
+
+        let mut buf = vec![0u8; 4096];
+        let btf_info = match sys::btf_obj_get_info_by_fd(btf_fd, &mut buf) {
+            Ok(info) => {
+                if info.btf_size > buf.len() as u32 {
+                    buf.resize(info.btf_size as usize, 0u8);
+                    let btf_info =
+                        sys::btf_obj_get_info_by_fd(btf_fd, &mut buf).map_err(|io_error| {
+                            ProgramError::SyscallError {
+                                call: "bpf_obj_get_info_by_fd".to_owned(),
+                                io_error,
+                            }
+                        })?;
+                    Ok(btf_info)
+                } else {
+                    Ok(info)
+                }
+            }
+            Err(io_error) => Err(ProgramError::SyscallError {
+                call: "bpf_obj_get_info_by_fd".to_owned(),
+                io_error,
+            }),
+        }?;
+
+        let btf_data = buf.as_slice();
+        let btf = Btf::parse(
+            &btf_data[0..btf_info.btf_size as usize],
+            Endianness::default(),
+        )
+        .map_err(ProgramError::Btf)?;
+
+        let btf_id = btf
+            .id_by_type_name_kind(func_name, BtfKind::Func)
+            .map_err(ProgramError::Btf)?;
+
+        self.data.attach_btf_obj_fd = Some(btf_fd as u32);
+        self.data.attach_prog_fd = Some(target_prog_fd);
+        self.data.attach_btf_id = Some(btf_id);
+        load_program(BPF_PROG_TYPE_EXT, &mut self.data)
+    }
+
+    /// Attaches the program to the given program
+    pub fn attach(&mut self) -> Result<LinkRef, ProgramError> {
+        let prog_fd = self.data.fd_or_err()?;
+        let target_fd = self.data.attach_prog_fd.ok_or(ProgramError::NotLoaded)?;
+        let btf_id = self.data.attach_btf_id.ok_or(ProgramError::NotLoaded)?;
+        let link_fd = bpf_link_create(prog_fd, target_fd, BPF_CGROUP_INET_INGRESS, Some(btf_id), 0)
```

Comment:
surely the attach type shouldn't be hardcoded here

Reply:
actually it does need to be set to `0` here, i'll add a comment as the BPF_GROUP_INET_INGRESS is misleading